refinery/tool_manager/models.py

```diff
@@ -793,8 +794,11 @@ def create_analysis_output_node_connections(self):
         """Create the AnalysisNodeConnection objects corresponding to the
         output Nodes (Derived Data) of a WorkflowTool launch
         """
-        exposed_workflow_outputs = self._get_exposed_galaxy_datasets()
-        for galaxy_dataset in self._get_galaxy_history_dataset_list():
+        exposed_dataset_list = self._get_galaxy_history_dataset_list()
+        exposed_workflow_outputs = self._get_exposed_galaxy_datasets(
+            exposed_dataset_list=exposed_dataset_list
+        )
```

Comment:
@hackdna these two functions are only every used here, so my question is similar to my other comment.  not being familiar enough with the system, i wonder if they should be removed and all their functionality be wrapped up into create_analysis_output_node_connections() ?  is there a reason to keep around these functions that specifically make the set of API calls/data filtering they make?  i aimed for maximizing speed with this PR rather than refactoring completely in this PR.

Reply:
I am not familiar with this part of the system either but I suspect the whole `tool_manager` app probably requires a re-write.

Reply:
I think that gets to the heart of why this issue has taken so long to resolve - I am walking a line between solving the bug at hand and refactoring large swaths of code, which was not what this was originally supposed to be.  We can do that, but I was always under the impression that was going to be the next step as it might take some time to do that.

Reply:
One of the reasons this code needs cleanup is that band-aid solutions were often used. Milestone 1.7.2 has just started, so might be a good time to refactor or rewrite.

Reply:
ok @hackdna i think then we should leave this alone in this pull request since it's not really relevant to the solution - i'll update the above create_analysis_output_node_connections() function to return the values mentioned above (and remove the database schema change).  my next pull request on this matter will then be the refactor; that way, i can separate the more cosmetic change from this, which is a bug-fix.  does that sound like a good plan?

Reply:
OK, as long as you follow up :)